Add Wi-Fi Keys API

adrianiainlam opened the pull request:
This is an incomplete first draft.

This branch also contains PR #25608 which may be superseded by PR #25734 but I have not had time to test it yet.

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, click the 'edited' button above_</sup>
Build number: 32
### Inputs:
#### Sources:
sdk-nrf: PR head: [f11b1f11fb841020345b11513908599702778d70](https://github.com/nrfconnect/sdk-nrf/pull/25745/commits/f11b1f11fb841020345b11513908599702778d70)
<details><summary> more details</summary>

sdk-nrf:
> PR head: [f11b1f11fb841020345b11513908599702778d70](https://github.com/nrfconnect/sdk-nrf/pull/25745/commits/f11b1f11fb841020345b11513908599702778d70)
> merge base: [8d95a2d1c21b4855ddb82d46b6c70155d2e3d3f3](https://github.com/nrfconnect/sdk-nrf/commit/8d95a2d1c21b4855ddb82d46b6c70155d2e3d3f3)
> target head (main): [8d95a2d1c21b4855ddb82d46b6c70155d2e3d3f3](https://github.com/nrfconnect/sdk-nrf/commit/8d95a2d1c21b4855ddb82d46b6c70155d2e3d3f3)
> [Diff](https://github.com/nrfconnect/sdk-nrf/compare/8d95a2d1c21b4855ddb82d46b6c70155d2e3d3f3..f11b1f11fb841020345b11513908599702778d70)</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 
|| ci-run-draft | _Run CI on draft pull requests_ | 

<details><summary> List of changed files detected by CI (20)</summary>

```
CODEOWNERS
drivers
│  ├── wifi
│  │  ├── nrf71
│  │  │  ├── CMakeLists.txt
│  │  │  ├── Kconfig
│  │  │  ├── src
│  │  │  │  │ wpa_supp_if.c
include
│  ├── wifi_kmu
│  │  │ wifi_kmu.h
lib
│  ├── CMakeLists.txt
│  ├── Kconfig
│  ├── wifi_kmu
│  │  ├── CMakeLists.txt
│  │  ├── Kconfig
│  │  ├── wifi_kmu.c
│  │  │ wifi_kmu_utils.c
modules
│  ├── trusted-firmware-m
│  │  ├── tfm_boards
│  │  │  │ CMakeLists.txt
tests
│  ├── lib
│  │  ├── wifi_kmu
│  │  │  ├── CMakeLists.txt
│  │  │  ├── prj.conf
│  │  │  ├── src
│  │  │  │  │ main.c
│  │  │  │ testcase.yaml
│  ├── wifi
│  │  ├── crypto
│  │  │  ├── CMakeLists.txt
│  │  │  ├── prj.conf
│  │  │  ├── src
│  │  │  │  │ main.c
│  │  │  │ testcase.yaml
```
</details>

### Outputs:

#### Toolchain
Version: 360200a5ee
Build docker image: docker-dtr.nordicsemi.no/sw-production/ncs-build:360200a5ee_a140d687bb

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :white_medium_square: Toolchain - Skipped: existing toolchain is used
- :white_check_mark: Build twister
   - sdk-nrf test count: 893
- :white_check_mark: Integration tests
   - :white_check_mark: test-fw-nrfconnect-nrf_crypto
   - :white_check_mark: test-fw-nrfconnect-tfm
   - :white_check_mark: test-sdk-find-my
   - :white_check_mark: test-sdk-wifi

<details><summary>Disabled integration tests</summary>

- 
   - test-fw-nrfconnect-nrf_lrcs_mosh
   - test-fw-nrfconnect-nrf_lrcs_positioning
   - desktop52_verification
   - test_ble_nrf_config
   - test-fw-nrfconnect-apps
   - test-fw-nrfconnect-apps_nrfdesktop
   - test-fw-nrfconnect-ble_mesh
   - test-fw-nrfconnect-ble_samples
   - test-fw-nrfconnect-fem
   - test-fw-nrfconnect-nfc
   - test-fw-nrfconnect-nrf-iot_libmodem-nrf
   - test-fw-nrfconnect-nrf-iot_lwm2m
   - test-fw-nrfconnect-nrf-iot_samples
   - test-fw-nrfconnect-nrf-iot_zephyr_lwm2m
   - test-fw-nrfconnect-rpc
   - test-fw-nrfconnect-rs
   - test-fw-nrfconnect-thread-main
   - test-low-level
   - test-sdk-audio
   - test-sdk-dfu
   - test-sdk-mcuboot
   - test-secdom-samples-public
</details>



*Note: This message is automatically posted and updated by the CI*

degjorva reviewed (changes requested):
Why is this added to nrf_security? it is just a wrapper crypto functionality that Wi-FI uses and as such should live in Wi-Fi. This is how most subsystems handle this (Bluetooth, Bootloaders etc).

krish2718 commented on subsys/net/lib/wifi_keys/Kconfig:
```diff
@@ -0,0 +1,21 @@
+
+# Configure WIFI_KEYS_LOG_LEVEL
+module = WIFI_KEYS
+module-dep = LOG
```
```suggestion
```

krish2718 commented on subsys/net/lib/wifi_keys/wifi_keys.c:
```diff
@@ -0,0 +1,250 @@
+	}
+
+	return ram_base + db_base + offset;
+}
```
Maybe use same names as LMAC for easier correlation?
```suggestion
#define WIFI_KEYS_RAM_BASE           0x28400000
#define WIFI_KEYS_VIF_KEY_DB_OFFSET  0x1000
#define WIFI_KEYS_VIF_MIC_LEN_PER_VIF  0x40
#define WIFI_KEYS_VIF_KEY_LEN_PER_VIF  0x80
#define WIFI_KEYS_VIF_MIC_LEN_PER_KEYID  0x10
#define WIFI_KEYS_VIF_KEY_LEN_PER_KEYID  0x20
#define WIFI_KEYS_PEER_MIC_LEN_16    0x10
#define WIFI_KEYS_PEER_UCST_KEY_LEN  0x20
#define WIFI_KEYS_PEER_BCST_KEY_LEN  0x20
#define WIFI_KEYS_PEER_KEY_TOTAL_LEN 0xf0
static uint32_t wifi_keys_get_key_start_addr(wifi_keys_key_type_t type, uint32_t db_id,
					     uint32_t key_index)
{
	const uint32_t ram_base = 0x28400000;
	uint32_t db_base;
	uint32_t offset;
	bool mic = wifi_keys_key_type_is_mic(type);

	if (type == VIF_ENC || type == VIF_MIC) {
		if (db_id >= 4) {
			return WIFI_KEYS_KEY_INDEX_INVALID;
		}
		db_base = 0x1000 + db_id * 0xC0;
		if (key_index >= 4) {
			return WIFI_KEYS_KEY_INDEX_INVALID;
		}
		if (mic) {
			offset = key_index * 0x10;
		} else {
			offset = 0x40 + key_index * 0x20;
		}
	} else { /* PEER */
		if (db_id >= 8) {
			return WIFI_KEYS_KEY_INDEX_INVALID;
		}
		db_base = db_id * 0xF0;
		if (type == PEER_UCST_ENC || type == PEER_UCST_MIC) {
			offset = mic ? 0x0 : 0x10;
		} else { /* PEER_BCST */
			if (key_index >= 4) {
				return WIFI_KEYS_KEY_INDEX_INVALID;
			}
			if (mic) {
				offset = 0x30 + key_index * 0x10;
			} else {
				offset = 0x70 + key_index * 0x20;
			}
		}
	}

	return ram_base + db_base + offset;
}
```

krish2718 reviewed (changes requested):
Left comments, with local fixes in https://github.com/krish2718/sdk-nrf/tree/kmu_71 security association works with AP.

krish2718 reviewed (approved):
Tested personal security on FPGA and it works fine.

nordicjm commented on subsys/nrf_security/src/drivers/kmu_push/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+#
+
+target_include_directories(psa_interface
+    INTERFACE
```
2 space indent for CMake

adrianiainlam replied:
Please direct comments for kmu_push to PR #27418, I am merely the user, not the author.

nordicjm commented on tests/subsys/nrf_security/kmu_push/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project("KMU push unit tests")
```
```suggestion
project(kmu_push)
```

nordicjm commented on subsys/net/lib/wifi_keys/Kconfig:
```diff
@@ -0,0 +1,17 @@
+config HAS_HW_NRF_WIFI_KEYS
+	def_bool SOC_SERIES_NRF71
+
+config WIFI_KEYS_DRIVER
```
if this is a driver, why is it in subsys?

adrianiainlam replied:
Right. I had it in nrf_security/src/drivers, was told to move it out. [I asked if subsys/net/lib is OK, and no-one answered.](https://github.com/nrfconnect/sdk-nrf/pull/25745#issuecomment-3739579804).

nordicjm replied:
what I mean is: this is not a driver, so this should not have driver in the Kconfig

nordicjm commented on tests/wifi/crypto/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(nrf_wifi_crypto)
+
+FILE(GLOB app_sources src/*.c)
```
do not use a glob for 1 file

nordicjm replied:
not fixed?

nordicjm commented on tests/wifi/crypto/Kconfig:
```diff
@@ -0,0 +1,6 @@
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+source "Kconfig.zephyr"
```
file to go

adrianiainlam commented on lib/wifi_kmu/wifi_kmu_utils.c:
My rationale for having these functions in a separate file is so that they can have different code owners - @frkv's team can own wifi_kmu.c and I can own this.
Is everyone OK with this approach? If so, I can add it to CODEOWNERS.

adrianiainlam replied:
Frank has transferred ownership to me. But I think it still makes sense to have two separate files - one for the actual KMU interaction, and one with more Wi-Fi specific address information etc.

krish2718 commented:
Verified STA mode WPA2-PSK and connection and ping are successful using the latest nrfx changes on a SoC FPGA.

nordicjm commented on include/wifi_kmu.h:
```diff
@@ -0,0 +1,62 @@
+ */
+
+#ifndef WIFI_KMU_H__
+#define WIFI_KMU_H__
```
move include to sub folder

nordicjm commented on lib/wifi_kmu/Kconfig:
```diff
@@ -0,0 +1,34 @@
+	help
+	  This is the lowest KMU key slot of pushed WI-FI keys.
+	  The default is set to the highest number of unreserved keys.
+
```
```suggestion
```

nordicjm commented on modules/trusted-firmware-m/tfm_boards/CMakeLists.txt:
```diff
@@ -115,6 +115,19 @@ if(${TFM_PARTITION_CRYPTO})
     endif()
   endif()
 
+  if(${CONFIG_WIFI_KMU})
```
why does this have `${}` ?

adrianiainlam replied:
I'm not the author, @frkv is, so I can't tell you why; but if you're saying it should not, then does that mean line 87 is wrong also?

adrianiainlam replied:
Not a fan of answering questions, I see.

nordicjm replied:
it is wrong

nordicjm commented on tests/lib/wifi_kmu/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(NONE)
+
+FILE(GLOB app_sources src/*.c)
```
no glob for 1 .c file

nordicjm commented on tests/lib/wifi_kmu/Kconfig.sysbuild:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+config PARTITION_MANAGER
```
file to go

nordicjm reviewed (changes requested):
I started looking and then saw existing issues were not fixed, then saw fixup commits and have abandoned the review - no fixup commits, squash them into the original commits, please re-request my review when this has been done and comments left have been addressed

adrianiainlam commented:
@nordicjm @degjorva Did you get notifications for the review requests?

nordicjm commented on lib/wifi_kmu/wifi_kmu_utils.c:
```diff
@@ -0,0 +1,183 @@
+#include "wifi_kmu/wifi_kmu.h"
+
+/* Wi-Fi Keys sizes */
+static const uint32_t MIC_KEY_LEN = 0x10;
```
coding style is lower snake case for variables, defines is upper case

nordicjm commented on tests/lib/wifi_kmu/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(NONE)
```
name is `wifi_kmu`

adrianiainlam commented on tests/wifi/crypto/src/main.c:
```diff
@@ -0,0 +1,81 @@
+
+	zassert_equal(err, 0);
+
+	*(volatile uint32_t *)0x48086C04 = 1; /* NRF_WIFICORE_RPUSYS->EDCPERIP.EDCGPIO1OUT */
+	while (*(volatile uint32_t *)0x48086C00 !=
+	       2) { /* NRF_WIFICORE_RPUSYS->EDCPERIP.EDCGPIO0OUT */
+	}
```
This is intended

adrianiainlam commented on lib/wifi_kmu/wifi_kmu.c:
```diff
@@ -0,0 +1,195 @@
+	if (slot_id + num_slots > KMU_RESERVED_SLOT_MIN) {
+		return -EINVAL;
+	}
+
+#if defined(CONFIG_BUILD_WITH_TFM)
+	/* WIFI KMU keys are transferred with IOCTL call */
+
+	return err;
+#else
```
This is intended as we don't support it yet.

adrianiainlam commented on lib/wifi_kmu/Kconfig:
```diff
@@ -0,0 +1,34 @@
+
+if NRF_WIFI_KMU
+
+config NRF_WIFI_KMU_SLOT_MIN
+	int "Lowest KMU slot for WI-FI keys in nRF71 series devices"
+	default 128
```
I disagree. This will mean that `KMU_RESERVED_SLOT_MIN` being 180 is defined in two separate places, leading to poor maintainability if e.g. the value needs to be changed in the future.

adrianiainlam commented on drivers/wifi/nrf71/src/wpa_supp_if.c:
```diff
@@ -20,8 +20,7 @@
 
 #ifdef CONFIG_NRF_WIFI_USE_KMU
-#include <psa/crypto.h>
-#include "wifi_keys.h"
+#include "wifi_kmu/wifi_kmu.h"
 #endif
```
Fixed in the previous unrelated update, CONFIG_NRF_WIFI_USE_KMU now depends on CONFIG_NRF_WIFI_KMU.

nordicjm commented on drivers/wifi/nrf71/src/wpa_supp_if.c:
```diff
@@ -1050,13 +1042,28 @@ static int wifi_import_key_to_crypto(unsigned int suite, const unsigned char *ke
-		LOG_ERR("%s: Failed to import key to PSA: %d", __func__, status);
+	wifi_kmu_key_reverse_byte_order_in_place(key_buf, max_size);
+
+	uint32_t dest_address = wifi_kmu_get_key_start_addr(type, db_id, crypto_key_index);
```
declare variables at top of scope

nordicjm commented on lib/wifi_kmu/wifi_kmu.c:
```diff
@@ -0,0 +1,200 @@
+		return -EINVAL;
+	}
+
+	size_t num_slots = DIV_ROUND_UP(key_size, KMU_KEYSLOTBYTES);
```
fix in whole PR

nordicjm commented on lib/wifi_kmu/wifi_kmu_utils.c:
```diff
@@ -0,0 +1,183 @@
+				    uint32_t key_length_bytes)
+{
+	/* Fix byte order of keys from wpa_supplicant */
+
```
```suggestion
```

nordicjm commented on lib/wifi_kmu/wifi_kmu_utils.c:
```diff
@@ -0,0 +1,183 @@
+int wifi_kmu_key_reverse_byte_order_in_place(void *buf, uint32_t key_length_bytes)
+{
+	/* Fix byte order of keys from wpa_supplicant - in place version */
+
```
```suggestion
```

nordicjm commented on modules/trusted-firmware-m/tfm_boards/CMakeLists.txt:
```diff
@@ -115,6 +115,19 @@ if(${TFM_PARTITION_CRYPTO})
     endif()
   endif()
 
+  if(${CONFIG_NRF_WIFI_KMU})
```
remove ${}

nordicjm commented on tests/wifi/crypto/src/main.c:
```diff
@@ -0,0 +1,81 @@
+#include <stdbool.h>
+#include <errno.h>
+
+#include "wifi_kmu/wifi_kmu.h"
```
all of these are the wrong type, use `<>`, this is a global include path, not local

degjorva commented on lib/wifi_kmu/wifi_kmu.c:
```diff
@@ -0,0 +1,200 @@
+#define KMU_RESERVED_SLOT_MIN (180)
+
+/* SICR page for KMU on nRF71 series devices */
+#define MRAM_CONFIGNVR_SICR_PAGE 3
```
Should this not be a nrfx define?

adrianiainlam replied:
I don't know. The same thing was added in cracen_psa_kmu.c in c01a4a92ac0.

degjorva reviewed (approved):
Looks good

adrianiainlam committed:
network: Wi-Fi: Add library for Wi-Fi KMU key pushing

-This library will be used with and without TF-M to allow pushing
 Wi-Fi keys to an external core by the use of the KMU
 (Key Management Unit).
-This library is only dependent on nrfx APIs to ensure it is
 directly usable in scope of TF-M.
-This adds unit test for the wifi_kmu library. It is only usable
 on nRF71 series devices. The tests build targets with and without
 TF-M to be debug-friendly
-Add Wi-Fi crypto utility functions and test for nRF71, including
 key address lookup, KMU slot rotation, and byte reordering for
 wpa_supplicant.

ref: NCSDK-36535, CAL-6362, WZN-10216

Co-authored-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>
Co-authored-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
Signed-off-by: Adrian Lam <adrian.lam@nordicsemi.no>
